docs/ql-documentation/learn-ql/javascript/flow-labels.rst

```diff
@@ -0,0 +1,391 @@
+
+Plain data-flow configurations implicitly use a single flow label "data", which indicates that a
+data value originated from a source. You can use the predicate ``DataFlow::FlowLabel::data()``,
+which returns this flow label, as a symbolic name for it.
+
+Taint-tracking configurations add a second flow label "taint" (``DataFlow::FlowLabel::taint()``),
+which is similar to "data", but includes values that have passed through non-value preserving steps
+such as string operations.
+
+Each of the three member predicates ``isSource``, ``isSink`` and
+``isAdditionalFlowStep``/``isAdditionalTaintStep`` has one version that uses the default flow
+labels, and one version that allows specifying custom flow labels through additional arguments.
+
+For ``isSource``, there is one additional argument specifying which flow label(s) should be
+associated with values originating from this source. If multiple flow labels are specified, each
+value is associated with `all` of them.
+
+For ``isSink``, the additional argument specifies which flow label(s) a value that flows into this
+source may be associated with. If multiple flow labels are specified, then any value that is
+associated with `at least one` of them will be considered by the configuration.
+
+For ``isAdditionalFlowStep`` there are two additional arguments ``predlbl`` and ``succlbl``, which
+allow flow steps to act as flow label transformers. If a value associated with ``predlbl`` arrives
+at the start node of the additional step, it is propagated to the end node and associated with
+``succlbl``. Of course, ``predlbl`` and ``succlbl`` may be the same, indicating that the flow step
+preserves this label. There can also be multiple values of ``succlbl`` for a single ``predlbl`` or
+vice versa.
+
+Note that if you do not restrict ``succlbl`` then it will be allowed to range over all flow labels.
+This may cause labels that were previously blocked on a path to reappear, which is not usually what
+you want.
+
+The flow label-aware version of ``isBarrier`` is called ``isLabeledBarrier``: unlike ``isBarrier``,
+which prevents any flow past the given node, it only blocks flow of values associated with one of
+the specified flow labels.
+
+Standard queries using flow labels
+----------------------------------
+
+Some of our standard security queries use flow labels. You can look at their implementation
+to get a feeling for how to use flow labels in practice.
+
+In particular, both of the examples mentioned in the section on limitations of basic data flow above
+are from standard security queries that use flow labels. The `Prototype pollution
+<https://lgtm.com/rules/1508857356317>`_ query uses two flow labels to distinguish completely
+tainted objects from partially tainted objects. The `Uncontrolled data used in path expression
+<https://lgtm.com/rules/1971530250>`_ query uses four flow labels to track whether a user-controlled
+string may be an absolute path and whether it may contain ``..`` components.
```

Comment:
Suggest adding a `What next?` section, as we do in the other tutorials:
For instance, you could just copy/paste this:
```
What next?
----------

-  Learn about the QL standard libraries used to write queries for JavaScript in :doc:`Introducing the Javacript libraries <introduce-libraries-js>`.
-  Find out more about QL in the `QL language handbook <https://help.semmle.com/QL/ql-handbook/index.html>`__ and `QL language specification <https://help.semmle.com/QL/ql-spec/language.html>`__.
-  Learn more about the query console in `Using the query console <https://lgtm.com/help/lgtm/using-query-console>`__.
```